sgv2-service-common/src/main/java/io/stargate/sgv2/common/grpc/DefaultGrpcSchema.java

```diff
@@ -0,0 +1,191 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.sgv2.common.grpc;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import io.grpc.Channel;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+import io.stargate.grpc.StargateBearerToken;
+import io.stargate.proto.QueryOuterClass.SchemaChange;
+import io.stargate.proto.Schema.CqlKeyspaceDescribe;
+import io.stargate.proto.Schema.DescribeKeyspaceQuery;
+import io.stargate.proto.Schema.GetSchemaNotificationsParams;
+import io.stargate.proto.Schema.SchemaNotification;
+import io.stargate.proto.StargateBridgeGrpc;
+import io.stargate.proto.StargateBridgeGrpc.StargateBridgeStub;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class DefaultGrpcSchema implements GrpcSchema {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultGrpcSchema.class);
+
+  private static final int KEYSPACE_CACHE_SIZE = 1000;
+  private static final Duration KEYSPACE_CACHE_TTL = Duration.ofMinutes(5);
+  private static final int DESCRIBE_KEYSPACE_TIMEOUT_SECONDS = 5;
+  private static final GetSchemaNotificationsParams GET_SCHEMA_NOTIFICATIONS_PARAMS =
+      GetSchemaNotificationsParams.newBuilder().build();
+
+  private final Channel channel;
+  private final StargateBearerToken adminCallCredentials;
+  // Each entry represents the result of the query (that may be complete or still in progress) to
+  // describe a given keyspace. No entry means we know nothing about the keyspace, the first client
+  // to ask for it will trigger the query. `CompletionStage<null>` means the keyspace does not
+  // exist. Schema change notifications invalidate the entry to force a refresh.
+  private final Cache<String, CompletionStage<CqlKeyspaceDescribe>> keyspaceCache;
+
+  DefaultGrpcSchema(Channel channel, String adminAuthToken) {
+    this.channel = channel;
+    this.adminCallCredentials = new StargateBearerToken(adminAuthToken);
+    this.keyspaceCache =
+        Caffeine.newBuilder()
+            .maximumSize(KEYSPACE_CACHE_SIZE)
+            .expireAfterAccess(KEYSPACE_CACHE_TTL)
+            .build();
+    registerChangeObserver();
+  }
+
+  @Override
+  public CompletionStage<CqlKeyspaceDescribe> getKeyspaceAsync(String keyspaceName) {
+    LOG.debug("getKeyspaceAsync({})", keyspaceName);
+    CompletionStage<CqlKeyspaceDescribe> existing = keyspaceCache.getIfPresent(keyspaceName);
+    if (existing != null) {
+      LOG.debug("Found entry in the cache, returning");
```

Comment:
Maybe add `keyspaceName` in message (otherwise not very useful log statement)

Reply:
It should be fairly easy to correlate with the log from line 70, if the logs contain thread names.
But I can also merge them together...